### PR TITLE
JAMES-3149 Reactify GetMessages

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java [hunk 1]
```diff
@@ -43,6 +45,14 @@
 
     List<MessageResult> getMessages(Collection<MessageId> messageIds, FetchGroup minimal, MailboxSession mailboxSession) throws MailboxException;
 
+    default Flux<MessageResult> getMessagesReactive(Collection<MessageId> messageIds, FetchGroup minimal, MailboxSession mailboxSession) {
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java:48]: not a Publisher?

File: mailbox/api/src/main/java/org/apache/james/mailbox/RightManager.java [hunk 2]
```diff
@@ -105,7 +108,9 @@
      *         {@code session.getUser()} is null.
      * @throws MailboxException in case of unknown mailbox or unsupported right
      */
-    MailboxACL.Rfc4314Rights myRights(MailboxId mailboxId, MailboxSession session) throws MailboxException;
+    Rfc4314Rights myRights(MailboxId mailboxId, MailboxSession session) throws MailboxException;
+
+    Mono<Rfc4314Rights> myRightsReactive(MailboxId mailboxId, MailboxSession session);
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/RightManager.java:113]: idem?

**Comment by chibenwa** [comment]
```
[8b487da8969d9002a8670bc0109c23bc7553b1c3] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project apache-james-mailbox-cassandra: Compilation failure
[8b487da8969d9002a8670bc0109c23bc7553b1c3] [ERROR] /james-project/mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraTestSystemFixture.java:[88,16] constructor StoreMessageIdManager in class org.apache.james.mailbox.store.StoreMessageIdManager cannot be applied to given types;
[8b487da8969d9002a8670bc0109c23bc7553b1c3] [ERROR]   required: org.apache.james.mailbox.MailboxManager,org.apache.james.mailbox.store.MailboxSessionMapperFactory,org.apache.james.mailbox.events.EventBus,org.apache.james.mailbox.quota.QuotaManager,org.apache.james.mailbox.quota.QuotaRootResolver,org.apache.james.mailbox.store.PreDeletionHooks
[8b487da8969d9002a8670bc0109c23bc7553b1c3] [ERROR]   found: org.apache.james.mailbox.cassandra.CassandraMailboxManager,org.apache.james.mailbox.cassandra.CassandraMailboxSessionMapperFactory,org.apache.james.mailbox.events.EventBus,org.apache.james.mailbox.cassandra.ids.CassandraMessageId.Factory,org.apache.james.mailbox.quota.QuotaManager,org.apache.james.mailbox.store.quota.DefaultUserQuotaRootResolver,org.apache.james.mailbox.store.PreDeletionHooks
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java [hunk 3]
```diff
@@ -45,7 +46,7 @@
 
     List<MessageResult> getMessages(Collection<MessageId> messageIds, FetchGroup minimal, MailboxSession mailboxSession) throws MailboxException;
 
-    default Flux<MessageResult> getMessagesReactive(Collection<MessageId> messageIds, FetchGroup minimal, MailboxSession mailboxSession) {
+    default Publisher<MessageResult> getMessagesReactive(Collection<MessageId> messageIds, FetchGroup minimal, MailboxSession mailboxSession) {
         try {
             return Flux.fromIterable(getMessages(messageIds, minimal, mailboxSession));
```
Comment (Arsnael) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java:51]: I know you try to take advantage of the default here, but then you still should not use `Flux` in api ?

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java:51]: As an implementation detail but not part of the exposed API

**Comment by chibenwa** [comment]
```
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR] Errors: 
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldAddMandatoryPropertiesWhenNotInPropertyList:283 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldFetchHtmlMessage:224 » IllegalState Grouped...
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldFetchMessages:193 » IllegalState GroupedFlu...
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldNotFailOnSingleMessageFailure:627 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldNotOverrideTextBodyWhenItIsThere:387 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnAllPropertiesWhenNoPropertyGiven:264 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnEmptyTextBodyAndHtmlBodyWhenThoseAreEmpty:353 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnFullViewWhenRequestedTextBody:556 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnHeaderViewWhenRequestedTo:591 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnHeadersFieldWhenSpecificHeadersRequestedInPropertyList:421 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnKeywordsForMessageFlags:664 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnKeywordsWithoutForwardedWhenForwardedUserFlagsMessages:765 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnKeywordsWithoutUnsupportedKeywordsForMessageFlags:723 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnMetadataWhenOnlyMailboxIds:521 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnOneMessageWhenMessageInSeveralMailboxes:484 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnOnlyMandatoryPropertiesOnEmptyPropertyList:247 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnPropertyFilterWhenFilteringHeadersRequested:447 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnTextBodyWhenBodyInPropertyListAndEmptyHtmlBody:302 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   GetMessagesMethodTest.processShouldReturnTextBodyWhenEmptyTextBodyAndNotEmptyHtmlBody:325 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFastViewFactoryTest.fromMessageIdsShouldKeepProcessingEvenWhenFetchingFail » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFastViewFactoryTest.fromMessageIdsShouldReturnAMessageWithComputedFastProperties » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFastViewFactoryTest.fromMessageIdsShouldReturnAMessageWithPropertiesComputedFromFullMessageWhenNotPreComputed » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFastViewFactoryTest.fromMessageIdsShouldReturnMessagesWithPreviews » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFullViewFactoryTest.fromMessageResultsShouldCombineKeywords:165 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageFullViewFactoryTest.fromMessageResultsShouldReturnCorrectView:134 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageHeaderViewFactoryTest.fromMessageResultsShouldCombineKeywords:109 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageHeaderViewFactoryTest.fromMessageResultsShouldReturnCorrectView:84 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageMetadataViewFactoryTest.fromMessageResultsShouldCombineKeywords:92 » IllegalState
[fb0444c084750f9abf9600d4ff42f494b37eec03] [ERROR]   MessageMetadataViewFactoryTest.fromMessageResultsShouldReturnCorrectView:76 » IllegalState
```

**Comment by chibenwa** [comment]
`java.lang.IllegalStateException: GroupedFlux allows only one Subscriber` ???

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 4]
```diff
@@ -152,6 +151,22 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
             .collect(Guavate.toImmutableList());
     }
 
+    @Override
+    public Flux<MessageResult> getMessagesReactive(Collection<MessageId> messageIds, FetchGroup fetchGroup, MailboxSession mailboxSession) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java:155]: any reason we don't implement getMessages with getMessagesReactive?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java [hunk 5]
```diff
@@ -34,10 +34,16 @@
 
 import com.google.common.collect.Multimap;
 
+import reactor.core.publisher.Flux;
+
 public interface MessageIdMapper {
 
     List<MailboxMessage> find(Collection<MessageId> messageIds, FetchType fetchType);
 
+    default Flux<MailboxMessage> findReactive(Collection<MessageId> messageIds, FetchType fetchType) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java:43]: Publisher?

Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java:43]: we are in mailbox-store ?

Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java:43]: true

**Comment by Arsnael** [comment]
```
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemoryVacationIntegrationTest>VacationIntegrationTest.jmapVacationShouldSendNotificationTwiceWhenVacationReset:252 » ConditionTimeout
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySetMessagesMethodCucumberTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySetMessagesMethodCucumberTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySetMessagesMethodCucumberTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySetMessagesMethodCucumberTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] [ERROR]   MemorySharingTest » NullPointer
```

**Comment by chibenwa** [comment]
```
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 09:19:55.774 [ERROR] o.a.j.j.d.m.m.v.MessageViewFactory - Can not convert MessageResults to Message for 1
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread reactor-http-epoll-6
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:77)
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at reactor.core.publisher.Mono.block(Mono.java:1663)
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at org.apache.james.jmap.draft.model.message.view.MessageFullViewFactory.retrieveProjection(MessageFullViewFactory.java:136)
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at org.apache.james.jmap.draft.model.message.view.MessageFullViewFactory.fromMetaDataWithContent(MessageFullViewFactory.java:102)
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at org.apache.james.jmap.draft.model.message.view.MessageFullViewFactory.fromMessageResults(MessageFullViewFactory.java:173)
[96f2ebc677029e0f93a1c4dacb2de980450db1ce] 	at org.apache.james.jmap.draft.model.message.view.MessageViewFactory$Helpers.lambda$toMessageViews$1(MessageViewFactory.java:125)
```

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/RequestHandler.java [hunk 6]
```diff
@@ -76,18 +76,20 @@ public RequestHandler(Set<Method> methods, JmapRequestParser jmapRequestParser,
...
-                        return error(request, generateInvalidArgumentError(e.getMessage()));
-                    }
-                };
+            try {
+                JmapRequest jmapRequest = jmapRequestParser.extractJmapRequest(request, method.requestType());
+                return method.process(jmapRequest, request.getMethodCallId(), mailboxSession)
+                    .onErrorResume(JmapFieldNotSupportedException.class, e -> errorNotImplemented(e, request));
+            } catch (IOException e) {
+                LOGGER.error("Error occured while parsing the request.", e);
+                if (e.getCause() instanceof JmapFieldNotSupportedException) {
```
Comment (Arsnael) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/RequestHandler.java:85]: Isn't `JmapFieldNotSupportedException` being caught below?

Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/RequestHandler.java:85]: We should wrap it's instanciation in a mono to simplify error handling here, for sure.

**Comment by chibenwa** [comment]
```
[6bc84e89a68b88aa1646a5c0b82750b115d98db9] [ERROR] Failed to execute goal on project event-sourcing-pojo: Could not resolve dependencies for project org.apache.james:event-sourcing-pojo:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus:codehaus-parent:pom:4 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/codehaus-parent/4/codehaus-parent-4.pom.part (No such file or directory) -> [Help 1]
```

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactory.java [hunk 7]
```diff
@@ -92,18 +92,19 @@ public MessageFullViewFactory(BlobManager blobManager, MessageContentExtractor m
         return Helpers.toMessageViews(messages, this::fromMessageResults);
     }
 
-    public MessageFullView fromMetaDataWithContent(MetaDataWithContent message) throws IOException {
+    public Mono<MessageFullView> fromMetaDataWithContent(MetaDataWithContent message) throws IOException {
         Message mimeMessage = Helpers.parse(message.getContent());
```
Comment (mbaechler) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactory.java:96]: these steps could be long, should be included into the mono, right?

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**